egui/src/widgets/plot/mod.rs

```diff
@@ -628,85 +658,24 @@ impl Prepared {
         }
 
         let interact_radius: f32 = 16.0;
-        let mut closest_value = None;
         let mut closest_item = None;
         let mut closest_dist_sq = interact_radius.powi(2);
         for item in items {
-            for value in &item.series().values {
-                let pos = transform.position_from_value(value);
-                let dist_sq = pointer.distance_sq(pos);
-                if dist_sq < closest_dist_sq {
-                    closest_dist_sq = dist_sq;
-                    closest_value = Some(value);
-                    closest_item = Some(item.name());
+            if let Some(closest) = item.closest(ui, pointer, transform, *show_x, *show_y) {
```

Comment:
This seems to create a bunch of shapes for all items and then throws all of them away except for maybe one set right? That is very inefficient.

Reply:
I just had an idea about that (commit 912b2f3)

Reply:
This looks better, I don't know enough about about the language to estimate the performance impact of creating potentially thousands of closures though. Did you consider splitting this into two functions, one to get the distance to the item and one to then get the hover shapes? Then you could do some `min_by` and then call the trait method to get the shapes just for that object.

Reply:
I considered the split with a simple `distance_square` and `hover_shapes` on `PlotItem`, but decided against because it required to calculate the distance twice. I think I can salvage it by putting an index instead of the closure, and having `fn hover_shapes(index: usize)` on `PlotItem`. I think at this point, I need to do some benchmark of the different ideas. I am not sure how to bench a gui app though, I saw some code in the repo. I will try to have something concrete later this week.

Reply:
Sorry for the delay, it took me some time to get the benchmarks somewhat clean. I implemented an alternative here https://github.com/niladic/egui/commit/24d6547c9f5dfb1a8e760ad622cd533fb16f32d3. Also, after trying to implement a non boxed closure here, I found out that I was mistaken on the feasibility. It is not possible without existential types (unstable feature).

I benched (code here https://github.com/niladic/egui/commit/fd8e3448ff052404748814160d3b96581e94f46c and https://github.com/niladic/egui/commit/ecf907487be35366483073666e8f0fcca8ffa444)
* code before this PR as baseline [feature/add-barchart-v0-bench](/niladic/egui/tree/feature/add-barchart-v0-bench)
* my first commit with all the allocs [feature/add-barchart-v1-bench](/niladic/egui/tree/feature/add-barchart-v1-bench)
* the closure implementation [feature/add-barchart-v2-bench](/niladic/egui/tree/feature/add-barchart-v2-bench)
* the alternative with indexes [feature/add-barchart-v3-bench](/niladic/egui/tree/feature/add-barchart-v3-bench)

Results are on this branch [feature/add-barchart-bench-results](/niladic/egui/tree/feature/add-barchart-bench-results). I tried to do it as clean as possible on my machine, but the benches are still a bit noisy. I would say they are really close to each other and there are no regressions in terms of perf. The most interesting bits are [the boxed closure vs before this PR as baseline](https://github.com/niladic/egui/blob/feature/add-barchart-bench-results/cargo-output-v2-closure-vs-v0.txt), [the alternative vs before this PR as baseline](https://github.com/niladic/egui/blob/feature/add-barchart-bench-results/cargo-output-v3-with-index-vs-v0.txt) and [the alternative vs the boxed closure as baseline](https://github.com/niladic/egui/blob/feature/add-barchart-bench-results/cargo-output-v3-with-index-vs-v2-with-closure.txt).

I also added some flamegraphs on what I think is high but somewhat realistic load:
* [boxed closure 100 barcharts of 100 bars](https://github.com/niladic/egui/blob/feature/add-barchart-bench-results/v2-with-closure_plot_100x100_bars.svg)
* [boxed closure 100 lines of 100 points](https://github.com/niladic/egui/blob/feature/add-barchart-bench-results/v2-with-closure_plot_100x100_lines.svg)
* [boxed closure 100x100 points](https://github.com/niladic/egui/blob/feature/add-barchart-bench-results/v2-with-closure_plot_100x100_points.svg)
* [alternative 100 barcharts of 100 bars](https://github.com/niladic/egui/blob/feature/add-barchart-bench-results/v3-with-index_plot_100x100_bars.svg)
* [alternative 100 lines of 100 points](https://github.com/niladic/egui/blob/feature/add-barchart-bench-results/v3-with-index_plot_100x100_lines.svg)
* [alternative 100x100 points](https://github.com/niladic/egui/blob/feature/add-barchart-bench-results/v3-with-index_plot_100x100_points.svg)

To sum up, since there are no regressions and both implementations are comparable in terms of perf, I have no strong preference. The alternative with indexes might be simpler and easier to work with down the line. Tell me if you have any preference or other idea, otherwise I will go with the alternative one.

Reply:
Hey thanks a lot for doing those benchmarks! I'll suggest we go with the alternative using indices.

Reply:
Hey thanks a lot for doing those benchmarks! I agree that we should go with the alternative using indices.